### PR TITLE
Skip result "skipped" on calculate overall result

### PR DESCRIPTION
**Pull request by cfconrad**
If a test module has the result "skipped", it is currently treated as
failed.
This change ignore test module with result "skipped" to calculate the
overall result.

Ticket: https://progress.opensuse.org/issues/43259

### CODE REVIEW

**Comment by Martchus** [comment]
But maybe it is time to add a unit test for this function.

**Comment by cfconrad** [comment]
Ok, I will have a look into it.

**Comment by cfconrad** [comment]
I have added some test, but I'm a bit confused as there is a `$job->skipped_module_count` which seems to count the test-module-result equals `none`.
And it's also handled differently like `none + ok => failed`.

**Comment by okurz** [comment]
I would be very careful with this change as to not loose why the behaviour was like this in before. So do you know what we would loose by this?

**Comment by coolo** [comment]
none is not the same as skipped. None means the module wasn't reported - i.e. it didn't even run. As such the job can't be passed if modules of it didn't run, hence the -> failed.

skipped is different - as it's not used by isotovideo, but only by the parser of external results and there skipped means "don't modify the result". If 100% are 'skipped', the result is either skipped or passed. 1% is 'none', we need it to 'fail' the job.

**Comment by cfconrad** [comment]
Ok I need to change the logic then for 100% skipped and make the result passed.

What about the `skipped_module_count` . I think it is used to produce this overview view.
Should we accumulate skipped tests to none or make a additional case?

**Comment by coolo** [comment]
for this use case we can put skipped behind the none icon :)

**Comment by Martchus** [comment]
@cfconrad So far `skipped_module_count` is updated if a job is inserted with no result at all or the result is explicitly `OpenQA::Jobs::Constants::NONE`. If you want to change the way it is counted the place to do it would be `OpenQA::Schema::Result::JobModules::insert` and `OpenQA::Schema::Result::JobModules::store_column`. But since @coolo already says that none and skipped are different it is likely dangerous to mix them here and instead add another column for counting that (e.g. `externally_skipped_module_count`).

**Comment by cfconrad** [comment]
Commit https://github.com/os-autoinst/openQA/pull/1985/commits/977c79632550775af8bedd15b8f66900e79d5763 should produce the accounting output.

I _wasn't_ really able to test it properly, as I didn't found the glue how to update the css, js-assets and pg-db in a "mojo" way. Can someone help me here?

A other thing I would like to discuss. The `none` results are called mostly in the code `none` beside db and `skipped_module_count`. And now we introduce a `externally_skipped_module_count`. I'm not happy with that naming, but don't know if it is possible/wonted to rename the db.
Any comments?

Thx clemens

File: assets/javascripts/tests.js [hunk 1]
```diff
@@ -196,6 +196,9 @@ function renderTestResult( data, type, row ) {
         if (data.none) {
             html += " " + data.none + "<i class='fa module_none fa-ban' title='modules skipped'></i>";
         }
+        if (data.none) {
```
Comment (Martchus) [inline assets/javascripts/tests.js]: That condition looks wrong.

**Comment by Martchus** [comment]
> but don't know if it is possible/wonted to rename the db

That would be possible, too. One would have to write a migration script to transfer the values from the old column to the new column.

> Can someone help me here?

You've already added the test I've asked for. You could add further check to it to also test whether the columns of the `$job` have been incremented as expected after the `$job->insert_module` calls.

If you also want to test the UI (and you should because I believe your JavaScript code contains a bug) you need to extend one of the selenium tests. To check whether the test result is now rendered as expected on the 'All tests' page the test code should likely go into `t/ui/01-list.t`. Since the fixtures obviously have no test where the new column is populated with a value other than zero you can just make something up (preferably only for this particular test by extending its `schema_hook` function).

File: assets/javascripts/tests.js (hunk 1 above)
Comment (cfconrad) [reply assets/javascripts/tests.js]: ups, true
thx

**Comment by cfconrad** [comment]
@Martchus thx for the hints! In the end I was able to test it. Output looks like this
restults_stats

`./script/upgradedb --prepare_upgrade` created extra changes like foreign keys. I manipulated the db-upgrade scripts by hand, so they only contains my changes.

**Comment by okurz** [comment]
@cfconrad Could you please revisit the documentation and extend it to cover your changes? I'm a bit lost on all the details of what could go wrong with "skipped". Maybe doc helps :)

**Comment by Martchus** [comment]
> I manipulated the db-upgrade scripts by hand, so they only contains my changes.

I doubt that this was necessary. Likely there's something wrong and the one who will create the next migration would run into the same problem again. I'll check out your branch locally and create the migration scripts again to see what you other changes it contains.

**Comment by Martchus** [comment]
I pushed an additional commit with the remaining changes you've been removing. See the commit message why those occurred. Now we have the migration nicely spitted into 2 separate commits (still one migration in the end). Since I'm already on your branch I can do some local testing by the way :-)

**Comment by Martchus** [comment]
I've tested the migration locally and the generated code doesn't work out of the box. But this should be fixed. Otherwise the next one which does a migration will run into the same problem again. So I'm looking into it.

**Comment by Martchus** [comment]
There was an issue with the foreign key it added, indeed. The issue was triggered by a relationship I've recently added. Apparently one can't use DBIx in the way I attempted - but it only breaks in the migration scripts so I didn't notice before. I've pushed a fix for that.

The remaining diff DBIx is producing now seems to acknowledge the CASCADE delete for developer sessions when a user has been deleted:  
```
           name: developer_sessions_fk_user_id
-          on_delete: ''
-          on_update: ''
+          on_delete: CASCADE
+          on_update: CASCADE
```

Maybe this test would work now: https://github.com/os-autoinst/openQA/blob/master/t/34-developer_mode-unit.t#L554

**Comment by cfconrad** [comment]
@okurz I can add some docs, but I'm not sure where.
What I found so far is http://open.qa/docs/#_jobs , but it  describes the results of a Job instead of a testmodule.
To make sure, I didn't overseen it, is there already a doc which I could extend?
Otherwise I think it's time to start it and maybe just place it below the test-job results.

**Comment by Martchus** [comment]
> To make sure, I didn't overseen it, is there already a doc which I could extend?

I'm not aware of any documentation for job module results.

**Comment by Martchus** [comment]
I'd like to merge this. @coolo @okurz Are you ok with this? I'm just asking because you initially gave feedback but not later on.

**Comment by pevik** [comment]
IMHO job title needs to be fixed as well, it's still red (https://openqa.suse.de/tests/2493518).

**Comment by coolo** [comment]
that page title *is* the overall state discussed here. openqa.suse.de has a deployment older than this.

**Pull request merged**